library/alloc/src/collections/vec_deque/mod.rs

```diff
@@ -2029,7 +2073,41 @@ impl<T, A: Allocator> VecDeque<T, A> {
     #[stable(feature = "deque_extras_15", since = "1.5.0")]
     #[track_caller]
     pub fn insert(&mut self, index: usize, value: T) {
-        assert!(index <= self.len(), "index out of bounds");
+        let _ = self.insert_mut(index, value);
+    }
+
+    /// Inserts an element at `index` within the deque, shifting all elements
+    /// with indices greater than or equal to `index` towards the back, and
+    /// returning a reference to it.
+    ///
+    /// Element at index 0 is the front of the queue.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `index` is strictly greater than the deque's length.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(push_mut)]
+    /// use std::collections::VecDeque;
+    ///
+    /// let mut vec_deque = VecDeque::from([1, 2, 3]);
+    ///
+    /// let x = vec_deque.insert_mut(1, 5);
+    /// *x += 7;
+    /// assert_eq!(vec_deque, &[1, 12, 2, 3]);
+    /// ```
+    #[unstable(feature = "push_mut", issue = "135974")]
+    #[track_caller]
+    #[must_use = "if you don't need a reference to the value, use VecDeque::insert instead"]
+    pub fn insert_mut(&mut self, index: usize, value: T) -> &mut T {
+        assert!(
+            index <= self.len,
+            "insertion index (is {index}) should be <= len (is {})",
+            self.len()
+        );
```

Comment:
Could you leave this assertion as it was? Previously:

```rust
assert!(index <= self.len(), "index out of bounds");
```

If you think another message would be more clear then that is fine to propose that in another PR (I don't think that's a bad idea). But please try to keep this one only about the new API without changing other behavior, I don't want to delay this by further iterating on assertion message wording.

Reply:
The formatting also has performance implications and should likely be moved to a `#[cold]` function (similar to `assert_failed`), if you decide to make this change in a separate PR.

Reply:
Reverted the message.